common/server/statshouse/add_metrics_batch.cpp

```diff
@@ -0,0 +1,61 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2021 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include "add_metrics_batch.h"
+#include <common/tl/methods/string.h>
+
+#include <utility>
+
+#include "common/tl/constants/statshouse.h"
+#include "common/tl/parse.h"
+#include "common/tl/store.h"
+
+namespace statshouse {
+
+constexpr int BUFFER_LEN = 1000;
+
+static int buffer[BUFFER_LEN];
+
+void Metric::tl_store() const {
+  tl_store_int(fields_mask);
+  tl_store_string(name.data(), name.length());
+  tl_store_int(0); // tags size
+  if (fields_mask & vk::tl::statshouse::metric_fields_mask::counter) {
+    tl_store_double(counter);
+  }
+  if (fields_mask & vk::tl::statshouse::metric_fields_mask::t) {
+    tl_store_long(t);
+  }
+  if (fields_mask & vk::tl::statshouse::metric_fields_mask::value) {
+    vk::tl::store_vector(this->value);
+  }
+  if (fields_mask & vk::tl::statshouse::metric_fields_mask::unique) {
+    vk::tl::store_vector(this->unique);
+  }
+  if (fields_mask & vk::tl::statshouse::metric_fields_mask::stop) {
+    vk::tl::store_vector(this->stop);
+  }
+}
+
+void add_value_metric(std::vector<Metric>& metrics, std::string name, double value, long time) {
+  constexpr int fields_mask = vk::tl::statshouse::metric_fields_mask::t | vk::tl::statshouse::metric_fields_mask::value;
+  metrics.push_back({.fields_mask = fields_mask, .name = std::move(name), .t = time, .value = {value}});
+}
+
+tl_query_payload make_tl_request_payload(const std::vector<Metric> &metrics) {
+  int req_buffer_pos = 0;
+  buffer[req_buffer_pos++] = TL_STATSHOUSE_ADD_METRICS_BATCH;
+  int fields_mask = vk::tl::statshouse::add_metrics_batch_fields_mask::ALL;
+  buffer[req_buffer_pos++] = fields_mask;
+  buffer[req_buffer_pos++] = metrics.size();
+  for (const Metric &metric : metrics) {
```

Comment:
Let's create separate structure `StatshouseAddMetricBatch` and implement `tl_store` for it to avoid manual buffer filling

Reply:
fixed